presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,396 @@
+                if (!(TRUE_LITERAL).equals(filter)) {
+                    planNode = new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return new JoinEnumerationResult(lookup.getCumulativeCost(planNode, session, symbolAllocator.getTypes()), Optional.of(planNode));
+            }
+            return chooseJoinOrder(nodes, outputSymbols, isCrossJoinPhase);
+        }
+
+        private static boolean isJoinEqualityCondition(Expression expression)
+        {
+            return expression instanceof ComparisonExpression
+                    && ((ComparisonExpression) expression).getType() == EQUAL
+                    && ((ComparisonExpression) expression).getLeft() instanceof SymbolReference
+                    && ((ComparisonExpression) expression).getRight() instanceof SymbolReference;
+        }
+
+        private static JoinNode.EquiJoinClause toEquiJoinClause(ComparisonExpression equality, Set<Symbol> leftSymbols)
+        {
+            Symbol leftSymbol = Symbol.from(equality.getLeft());
+            Symbol rightSymbol = Symbol.from(equality.getRight());
+            JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(leftSymbol, rightSymbol);
+            return leftSymbols.contains(leftSymbol) ? equiJoinClause : equiJoinClause.flip();
+        }
+
+        private JoinEnumerationResult setJoinNodeProperties(JoinNode joinNode)
+        {
+            List<JoinEnumerationResult> possibleJoinNodes = new ArrayList<>();
+            FeaturesConfig.JoinDistributionType joinDistributionType = getJoinDistributionType(session);
+            if (joinDistributionType.canRepartition() && !joinNode.isCrossJoin()) {
+                JoinNode node = joinNode.withDistributionType(PARTITIONED);
+                possibleJoinNodes.add(new JoinEnumerationResult(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes()), Optional.of(node)));
+                node = node.flipChildren();
+                possibleJoinNodes.add(new JoinEnumerationResult(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes()), Optional.of(node)));
+            }
+            if (joinDistributionType.canReplicate()) {
+                JoinNode node = joinNode.withDistributionType(REPLICATED);
+                possibleJoinNodes.add(new JoinEnumerationResult(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes()), Optional.of(node)));
+                node = node.flipChildren();
+                possibleJoinNodes.add(new JoinEnumerationResult(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes()), Optional.of(node)));
+            }
+
+            return resultOrdering.min(possibleJoinNodes);
+        }
+    }
+
+    @VisibleForTesting
+    static class JoinEnumerationResult
+    {
+        static final JoinEnumerationResult UNKNOWN_COST_RESULT = new JoinEnumerationResult(UNKNOWN_COST, Optional.empty());
```

Comment:
I am not sure about that. Unknown cost breaks computation, while infinity allows for checking other partitions.